### ISSUE TITLE
Proposal: Improve Dashboard performance - rewrite

### ISSUE THREAD

**Issue by arvindsv** [gocd/gocd#1841]
Had to write this quickly, because @xli is working on [changing the dashboard](https://github.com/gocd/gocd/issues/1595#issuecomment-174402136), and I wanted to provide an option. :) JK. I don't expect this to be picked up right now. It is a proposal for a rewrite, albeit a clean one which can stay side-by-side with the current dashboard while it is in progress. But, it's good to have this out of my head and on digital-paper.

Broadly, the dashboard related code starts off [here](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/controllers/pipelines_controller.rb#L37), then [here](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/controllers/pipelines_controller.rb#L90-L94) and then [here](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/src/com/thoughtworks/go/server/service/PipelineHistoryService.java#L449). After the information about the pipelines is loaded, it goes through a set of ERB templates to create a [full dashboard page](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/views/pipelines/index.html.erb) and then a ["JSON"](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/views/pipelines/index.json.erb). I say JSON in quotes because it has HTML in it and is not really a JSON. Just in structure, not in spirit. Finally, there's a view cache layer [here](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/views/pipelines/_pipeline_group.html.erb#L11) and [here](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/views/pipelines/index.json.erb#L9).

The current problems with this approach are:
- The `PipelineHistoryService#allPipelineInstances` method creates un-cacheable `PipelineGroupModels` objects, which have user-specific information in them. Things like whether a user has permissions to trigger a pipeline or stage, whether the user can unlock the pipeline, etc.
- The ERBs are just wrong. They're too complex and don't produce proper JSONs. The view cache is not really needed. It's a per-user-per-pipeline-instance cache which is too fragile and gets changed too often. When a stage state changes, for instance, it's useless and has to be re-cached. Also, it is dependent on users. If a user has operate permission on a pipeline and another user doesn't, there will be two instances of this cached fragment.
- The view fragments in the cache do not get evicted quickly. Because of the way Rails' fragment caching works and how ehcache is setup, these fragments will stay in cache for a long, long time. Till 100,000 items are in the ehcache. Then, they might start getting evicted.
#### Proposal:

~~Move the PipelineInstanceModel from~~ Create a new PipelineInstanceModel class which goes from:

![pim_before](https://cloud.githubusercontent.com/assets/172235/12586928/3effb9be-c420-11e5-8242-0c0ea88a9afe.JPG)

to something like this:

![pim_after](https://cloud.githubusercontent.com/assets/172235/12586931/43d5bb96-c420-11e5-8dc2-8515e5739430.JPG)

This new instance model is cacheable, since it has no information specific to the current user.
##### Invalidation of this object:

There should only be 3 reasons for this object to be invalidated. Config change, job status change and stage status change. If one of these changes, invalidate the corresponding item from the cache. This is _exactly_ how [CCTray is implemented](https://github.com/gocd/gocd/blob/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/src/com/thoughtworks/go/domain/cctray/CcTrayActivityListener.java) and it scales pretty well.
##### UI layer:

I propose throwing away the whole ERB and view cache and giving us a clean slate to work on. Ideally GoCD dashboard is client-side and uses virtual dom diffing (Mithril). This is where that would really shine. Virtual dom diffs along with an incremental JSON update is what will really provide the performance benefit.

My proposal for the JSON is something like this:

```
{
  "etag": 1234,
  "pipeline_groups": [
    {
      "name": "pipeline_group1",
      "pipelines": ["pipeline1", "pipeline2"]
    },
    {
      "name": "pipeline_group2",
      "pipelines": ["pipeline3", "pipeline4"]
    }
  ],
  "pipelines": [
    {
      "name": "pipeline1",
      "can_lock": true,
      "can_operate": true,
      "build_cause": ...
      ...
    },
    {
      "name": "pipeline2",
      "can_lock": false,
      "can_operate": true,
      "build_cause": ...
      ...
    },
    {
      "name": "pipeline3",
      "can_lock": false,
      "can_operate": true,
      "build_cause": ...
      ...
    },
    {
      "name": "pipeline3",
      "can_lock": false,
      "can_operate": true,
      "build_cause": ...
      ...
    }
  ]
```

Provide that "etag" back to the server (it can be in a header, I don't care), and it should give you something like:

```
{
  "etag": 1256,
  "pipeline_groups": [
    {
      "name": "pipeline_group1",
      "pipelines": ["pipeline2", "pipeline1"]
    },
    {
      "name": "pipeline_group2",
      "pipelines": ["pipeline4", "pipeline5"]
    }
  ],
  "pipelines": [
    {
      "name": "pipeline1",
      "can_lock": false,
      "can_operate": false,
      "build_cause": ...
      ...
    },
    {
      "name": "pipeline5",
      "can_lock": true,
      "can_operate": true,
      "build_cause": ...
      ...
    }
  ]
```

As you can see, this incremental update has handled a pipeline delete ("pipeline3"), a re-order of pipelines ("pipeline1" and "pipeline2") and an add of a pipeline ("pipeline5"). It has returned a new etag to use. As long as that etag is used properly in the cached `PipelineInstanceModel` object, then it's easy to find objects in the cache newer than a provided value.

Of course, this implies a rewrite of the dashboard UI using this JSON. I think we can rewrite it so that a job itself can be reached from the dashboard without drilling down using two or three clicks. It should be such that you can go from seeing a failing pipeline on the dashboard to the corresponding console output of the failing jobs in one click.

**Comment by xli** [comment]
What is "It should be such that you can go from seeing a failing pipeline on the dashboard to the corresponding console output of the failing jobs in one click."? 

@wpc and I talked about live update dashboard by websocket some days ago. He think maybe react.js provides better infrastructure for us to build better UI.
I think it is good idea, if we'd like to rewrite UI.

**Comment by arvindsv** [comment]
I mentioned mithril because the new config UI is being written using that and [msx](https://github.com/insin/msx) rather than react+jsx. @zabil and others did some spikes and chose it. React is a question more for the others on the team. My view is that I'm already quite hesitant seeing the complexity of these new frameworks. I'd rather not bring another one into the mix, especially since mithril and react are not very very different, given what we want.

I think that most people who're come to see a pipeline when it fails want to reach the failing console log or logs quickly. I've heard people say "Jenkins provides a link to the failing console log directly". What they don't realize is that a Jenkins job is very granular. It's very close to what is happening. A GoCD pipeline is two levels of abstraction above that: stage and then pipeline. They're actually seeing an aggregated view when a pipeline fails.

However, what we don't do well is to provide a way to drill down to the failing job quickly. I think the dashboard should do that. Complications are that there can be multiple jobs which fail in a stage. So, there's some UX thought needed there, but I don't see why we can't bring job level data up to that view and make sure that a user can easily reach a failing job without clicking on the stage and then the failing job.

**Comment by iliasbartolini** [comment]
Hi, adding only a feedback on this:

> I think that most people who come to see a pipeline when it fails want to reach
> the failing console log or logs quickly. I've heard people say "Jenkins provides 
> a link to the failing console log directly".

Having used GoCD for years and introduced it to dozens of users, customers and colleagues I can't stress enough how many times I heard this.
Putting a little bit of empathy in my pocket and a UX hat I'd like to emphasize that usually when a build is "red" and someone is trying to fix it is one of the key "emotional" moments in a user journey experience. 

Think: "Something is broken, may affect others and users, I want to fix it quickly"... and if we can avoid  one or both of those two extra clicks would be greatly appreciated. Many users get lost in where to look at in those two extra pages and in that moment they just want to go straight to the logs that are showing the error.
If you improve this, I'm sure GoCD will be a little bit more loved and users happy :)

**Comment by xli** [comment]
@arvindsv I didn't know about msx in Go repository. I don't see any reason to introduce other framework if we're moving to a new one. No wonder I see new javascripts stuff :) I thought js files under app/assets/javascripts are good examples.

**Comment by arvindsv** [comment]
> If you improve this, I'm sure GoCD will be a little bit more loved and users happy :)

:) I guess so too.

We should involve @naveenbhaskar and get him thinking about the UX when someone decides to pick this up.

**Comment by arvindsv** [comment]
@xli: Yeah, things are changing in the JS part of the codebase. Any time I think about the new frameworks, [this section](https://medium.com/@wob/the-sad-state-of-web-development-1603a861d29f#.uoc274zaa) comes to mind, though. :) Especially the hamster dump part of it.

**Comment by ketan** [comment]
> No wonder I see new javascripts stuff :) I thought js files under app/assets/javascripts are good examples.

The new javascript starts [here](https://github.com/gocd/gocd/tree/75f21bf7bebddc15fadf65651b709dbb4961ddbe/server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs) You'll see things being loaded via `require.js` and an application.js for just the single page app.

Should you need another piece of JS — just create another `application.js` like thing for your needs.

Also — new css is [here](https://github.com/gocd/gocd/tree/master/server/webapp/WEB-INF/rails.new/app/assets/stylesheets/pipeline_configs) the idea is to re-use as much from zurb foundation.

The css is currently used by the pipeline config page — the idea is to create multiple CSS files for each separate page (and reduce duplication using some `framework.css` that contains framework code like reset, foundation, base fonts etc.) — mostly to improve page load performance.

**Comment by xli** [comment]
I'm curious about how many stages only have one job.
So I spent some time write script https://gist.github.com/xli/1430de60ba5feee7aeb6 to get stats from Go server.

Here is 2 samples:

* Mingle team's Go server (Go01): 
    * pipelineCount: 53
    * stageCount: 140
    * stageHasOneJobCount: 102
    * stageHasMultipleJobCount: 38
* build.go.cd: 
    * pipelineCount: 15
    * stageCount: 30
    * stageHasOneJobCount: 19
    * stageHasMultipleJobCount: 11

More than 60% stages only have one job.
So I think even we only improve one job stage's case, lots of users will get benefit.

**Comment by arvindsv** [comment]
Very good point. Yes, I agree.

**Comment by rajiesh** [comment]
I think this issue is closed after 18.3.0 release with new dashboard. All issues with new dashboard and its performance and also all the design improvements can be tracked in separate issues, so closing this.
